frontends/parsers/parserDriver.cpp

```diff
@@ -0,0 +1,238 @@
+#include <boost/format.hpp>
+
+#include <cerrno>
+#include <cstdio>
+#include <iostream>
+#include <sstream>
+
+#include "frontends/common/constantFolding.h"
+#include "frontends/parsers/p4/p4lexer.hpp"
+#include "frontends/parsers/p4/p4parser.hpp"
+#include "frontends/parsers/v1/v1lexer.hpp"
+#include "frontends/parsers/v1/v1parser.hpp"
+#include "lib/error.h"
+
+
+#ifdef HAVE_LIBBOOST_IOSTREAMS
+
+#include <boost/iostreams/device/file_descriptor.hpp>
+#include <boost/iostreams/stream.hpp>
+
+namespace {
+
+/// A RAII helper class that provides an istream wrapper for a stdio FILE*. This
+/// is the efficient implementation for users with boost::iostreams installed.
+struct AutoStdioInputStream {
+    explicit AutoStdioInputStream(FILE* in)
+        : source(fileno(in), boost::iostreams::never_close_handle)
+        , buffer(source)
+        , stream(&buffer)
+    { }
+
+    std::istream& get() { return stream; }
+
+ private:
+    AutoStdioInputStream(const AutoStdioInputStream&) = delete;
+    AutoStdioInputStream(AutoStdioInputStream&&) = delete;
+
+    boost::iostreams::file_descriptor_source source;
+    boost::iostreams::stream_buffer<boost::iostreams::file_descriptor_source> buffer;
+    std::istream stream;
+};
+
+}  // anonymous namespace
+
+#else
+
+/// A RAII helper class that provides an istream wrapper for a stdio FILE*. This
+/// is an inefficient fallback implementation.
+struct AutoStdioInputStream {
+    explicit AutoStdioInputStream(FILE* in) {
+        char buffer[512];
+        while (fgets(buffer, sizeof(buffer), in))
+            stream << buffer;
+    }
+
+    std::istream& get() { return stream; }
+
+ private:
+    std::stringstream stream;
+};
+
+#endif
+
+
+namespace P4 {
+
+AbstractParserDriver::~AbstractParserDriver() { }
+
+void AbstractParserDriver::onReadToken(const char* text) {
+    auto posBeforeToken = Util::InputSources::instance->getCurrentPosition();
+    Util::InputSources::instance->appendText(text);
+    auto posAfterToken = Util::InputSources::instance->getCurrentPosition();
+    yylloc = Util::SourceInfo(posBeforeToken, posAfterToken);
+}
+
+void AbstractParserDriver::onReadLineNumber(const char* text) {
+    char* last;
+    errno = 0;
+    lineDirectiveLine = strtol(text, &last, 10);
+    if (errno != 0 ||
+        // we have not parsed the complete string
+        strlen(last) != 0)
+        ErrorReporter::instance.parser_error("Error parsing line number %s", text);
+}
+
+void AbstractParserDriver::onReadFileName(const char* text) {
+    lineDirectiveFile = cstring(text);
+    Util::InputSources::instance->mapLine(lineDirectiveFile, lineDirectiveLine);
+}
+
+void AbstractParserDriver::onReadIdentifier(cstring id) {
+    lastIdentifier = id;
+}
+
+void AbstractParserDriver::onParseError(const Util::SourceInfo& location,
+                                        const std::string& message) {
+    static const std::string unexpectedIdentifierError =
+        "syntax error, unexpected IDENTIFIER";
```

Comment:
This is actually present in the old code too, although the implementation is now different.